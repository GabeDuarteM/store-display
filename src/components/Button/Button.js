import React from 'react'

import PropTypes from 'prop-types'
import injectSheet from 'react-jss'
import cn from 'classnames'

export const styles = {
  button: {
    cursor: ({ disabled }) => (disabled ? 'not-allowed' : 'pointer'),
    width: ({ width }) => width,
    height: 50,
    backgroundColor: '#6e8be3',
    userSelect: 'none',
    color: '#ffffff',
    fontSize: 14,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 3,
    border: 0,
  },
}

const Button = ({
  classes,
  className,
  children,
  onClick,
  disabled,
  ...rest
}) => (
  <button
    className={cn(classes.button, className)}
    onClick={onClick}
    disabled={disabled}
    {...rest}
  >
    {children}
  </button>
)

export const propTypes = {
  /** @ignore */
  classes: PropTypes.object.isRequired,
  /** The content of the button */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf([PropTypes.node]),
    PropTypes.node,
  ]).isRequired,
  /** The click handler of the button. */
  onClick: PropTypes.func.isRequired,
  /** If present, it will be added to the root component */
  className: PropTypes.string,
  /** The width of the component */
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]), // eslint-disable-line react/no-unused-prop-types
  /** Prop that defines if the button is disabled or not */
  disabled: PropTypes.bool,
}

export const defaultProps = {
  width: 280,
  disabled: false,
  className: null,
}

Button.propTypes = propTypes
Button.defaultProps = defaultProps

export default injectSheet(styles)(Button)
