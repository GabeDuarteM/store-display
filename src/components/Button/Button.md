```js
<Button onClick={() => {}}>Load more</Button>
```

```js
const loading = require('../../assets/loading.svg')
;<div>
  <Button disabled onClick={() => {}}>
    <SvgIcon width={35} src={loading} />
  </Button>
</div>
```
