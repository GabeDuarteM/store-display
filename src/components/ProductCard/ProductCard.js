import React from 'react'

import PropTypes from 'prop-types'
import injectSheet from 'react-jss'
import cn from 'classnames'

import SvgIcon from '../SvgIcon/SvgIcon'
import heart from '../../assets/heart.svg'
import heartO from '../../assets/heart-o.svg'

export const styles = {
  productCard: {
    width: 280,
    height: 344,
    background: '#ffffff',
    overflow: 'hidden',
    transition: 'box-shadow 0.3s ease-in-out',
    '&:hover': {
      boxShadow: '0px 30px 50px rgba(0,0,0,0.15)',
    },
    '&:hover $wishlistButton': {
      visibility: 'visible !important',
    },
  },
  img: {
    backgroundImage: ({ imgUrl }) => `url(${imgUrl})`,
    backgroundPosition: 'center',
    backgroundSize: 'contain',
    backgroundRepeat: 'no-repeat',
    width: 260,
    height: 260,
    display: 'flex',
    justifyContent: 'space-between',
  },
  imgContainer: {
    width: 280,
    height: 280,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  wishlistButton: {
    visibility: 'hidden',
    cursor: 'pointer',
    transition: 'transform 180ms',
    transitionTimingFunction: 'cubic-bezier(.2,.7,.8,.7)',
    marginTop: 10,
    marginLeft: 8,
    '&:hover': {
      transform: 'scale(1.2,1.2)',
    },
    '&.wishlisted': {
      visibility: 'visible !important',
    },
  },
  newProduct: {
    width: 90,
    color: '#d0011b',
    height: 90,
    fontSize: 12,
    background: '#fff2f2',
    fontWeight: 500,
    display: 'flex',
    alignItems: 'flex-end',
    justifyContent: 'center',
    transform: 'rotate(45deg)',
    margin: -61,
    paddingBottom: 6,
    boxSizing: 'border-box',
    userSelect: 'none',
  },
  description: {
    display: 'flex',
    justifyContent: 'space-between',
    padding: 14,
  },
  brand: {
    color: '#585858',
    fontFamily: 'Montserrat',
    fontSize: 14.0,
    fontWeight: 500,
    textTransform: 'uppercase',
  },
  name: {
    color: '#646464',
    fontSize: 12.0,
    fontWeight: 300,
    width: 160,
    whiteSpace: 'nowrap',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    marginTop: 4,
    height: 16,
  },
  price: {
    color: '#585858',
    fontSize: 14.0,
    fontWeight: 400,
    marginRight: 6,
  },
}

const ProductCard = ({
  classes,
  brand,
  name,
  price,
  id,
  currency,
  isWishlisted,
  toggleIsWishlisted,
  isNew,
  className,
}) => (
  <div className={cn(classes.productCard, className)}>
    <div className={classes.imgContainer}>
      <div className={classes.img}>
        <SvgIcon
          className={cn(classes.wishlistButton, { wishlisted: isWishlisted })}
          onClick={() => toggleIsWishlisted(id)}
          width={20}
          height={20}
          color={`#${isWishlisted ? 'ff5454' : '808080'}`}
          src={isWishlisted ? heart : heartO}
        />
        {isNew && <div className={classes.newProduct}>New</div>}
      </div>
    </div>
    <div className={classes.description}>
      <div>
        <div className={classes.brand}>{brand}</div>
        <div className={classes.name}>{name}</div>
      </div>
      <div className={classes.price}>
        {currency}
        {Intl.NumberFormat('pt-BR', { minimumFractionDigits: 2 }).format(price)}
      </div>
    </div>
  </div>
)

export const propTypes = {
  /** @ignore */
  classes: PropTypes.object.isRequired,
  /** The url of the image to be displayed */
  imgUrl: PropTypes.string.isRequired, // eslint-disable-line react/no-unused-prop-types
  /** The brand of the product */
  brand: PropTypes.string.isRequired,
  /** The name of the product */
  name: PropTypes.string.isRequired,
  /** The price of the product */
  price: PropTypes.number.isRequired,
  /** The id of the product */
  id: PropTypes.string.isRequired,
  /** Gets called when the user clicks the wishlist button */
  toggleIsWishlisted: PropTypes.func.isRequired,
  /** The currency of the product's price */
  currency: PropTypes.string,
  /** Prop that define if the product is wishlisted or not */
  isWishlisted: PropTypes.bool,
  /** Prop that define if the product is new or not */
  isNew: PropTypes.bool,
  /** custom className to be applied on the root element */
  className: PropTypes.string,
}

export const defaultProps = {
  currency: 'R$',
  isWishlisted: false,
  isNew: false,
  className: null,
}

ProductCard.propTypes = propTypes
ProductCard.defaultProps = defaultProps

export default injectSheet(styles)(ProductCard)
