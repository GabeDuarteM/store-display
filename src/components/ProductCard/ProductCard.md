```js
initialState = {
  isWishlisted: false,
}
;<ProductCard
  imgUrl="https://images.eastbay.com/pi/18837091/zoom/nike-legend-2.0-long-sleeve-t-shirt-mens"
  brand="Nike"
  id="5a5180507dceaccd20a314e0"
  name="Legend 2.0 Long Sleeve"
  price={24.99}
  isWishlisted={state.isWishlisted}
  isNew
  toggleIsWishlisted={() => setState({ isWishlisted: !state.isWishlisted })}
/>
```
