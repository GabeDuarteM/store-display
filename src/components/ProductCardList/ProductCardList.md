```js
<ProductCardList
  products={[
    {
      id: '5a5180507dceaccd20a314e0',
      imgUrl: '//images.arcteryx.com/F17/450x625/Cerium-LT-Jacket-Rooibos.png',
      gender: 'woman',
      rating: 9,
      price: 96.44,
      isNew: true,
      name: 'Aguirrewatson Ritakirkland',
      brand: 'Mazine',
    },
    {
      id: '5a518050fbb2da0a6d228cf9',
      imgUrl:
        '//cdn.shopify.com/s/files/1/1104/4168/products/Allbirds_W_Wool_Runner_Kotare_GREY_ANGLE_900x900.png?v=1514576081',
      gender: 'man',
      rating: 8,
      price: 105.7,
      isNew: false,
      name: 'Arlineschneider Loriemartinez',
      brand: 'NIKE',
    },
    {
      id: '5a51804f840a30ab47a631a2',
      imgUrl: '//www.kicksusa.com/media/wysiwyg/brands/nike/nike-men.png',
      gender: 'woman',
      rating: 10,
      price: 27.59,
      isNew: true,
      name: 'Candacechase Blackturner',
      brand: 'NIKE',
    },
    {
      id: '5a518050bbbe951b1e561fed',
      imgUrl:
        'http://shop.autismspeaks.org/Shared/images/Product/AUT/AT17012.png',
      gender: 'woman',
      rating: 4,
      price: 235.1,
      isNew: false,
      name: 'Burkemejia Cochranbray',
      brand: 'Urban Classics',
    },
    {
      id: '5a51805085f01f4226367dfd',
      imgUrl:
        'http://cdn.shopify.com/s/files/1/1378/9837/products/lets-love-tshirt_grande.png?v=1509580185',
      gender: 'man',
      rating: 2,
      price: 257.36,
      isNew: true,
      name: 'Fitzgeraldwallace Agnesmarks',
      brand: 'Urban Classics',
    },
    {
      id: '5a518050b6fb69f117f82d7e',
      imgUrl:
        '//cdn.shopify.com/s/files/1/1734/3501/products/CODWW2MLTRYSHRT-OLV-S_1024x1024.png?v=1507908119',
      gender: 'man',
      rating: 3,
      price: 270.8,
      isNew: false,
      name: 'Celiamoss Dominiquerosales',
      brand: 'Urban Classics',
    },
    {
      id: '5a518050864901ecde5be490',
      imgUrl:
        'https://www.swazi.co.nz/wp-content/uploads/2014/11/Micro-Pants-Olive-PMP.png',
      gender: 'woman',
      rating: 9,
      price: 119.6,
      isNew: false,
      name: 'Davidsonblackwell Fowlerbenson',
      brand: 'NIKE',
    },
    {
      id: '5a518050e421e659c6e24c46',
      imgUrl: 'https://tru-spec-cdn.azureedge.net/media/1175/1060f.png',
      gender: 'man',
      rating: 10,
      price: 282.1,
      isNew: true,
      name: 'Perkinsjames Clinegalloway',
      brand: 'VERO MODA',
    },
    {
      id: '5a51805044d1030570bcab76',
      imgUrl: 'https://tru-spec-cdn.azureedge.net/media/1025/1436f.png',
      gender: 'woman',
      rating: 4,
      price: 296.55,
      isNew: false,
      name: 'Henriettagillespie Dellavargas',
      brand: 'NIKE',
    },
  ]}
  handleClickLoadMore={() => {}}
  toggleIsWishlisted={() => {}}
  isLastPage={false}
/>
```
