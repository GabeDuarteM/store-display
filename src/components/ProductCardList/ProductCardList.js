import React from 'react'

import PropTypes from 'prop-types'
import injectSheet from 'react-jss'
import cn from 'classnames'

import ProductCard from '../ProductCard'
import Button from '../Button'

export const styles = {
  productCardList: {
    display: 'flex',
    flexWrap: 'wrap',
    padding: '15px 0',
    margin: '0 47px',
    alignItems: 'center',
    justifyContent: 'center',
  },
  productCard: {
    flex: 'none',
    margin: 15,
  },
  loadMoreButtonRoot: {
    width: '100%',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
  loadMoreButton: {
    margin: 25,
  },
}

const ProductCardList = ({
  className,
  classes,
  products,
  toggleIsWishlisted,
  handleClickLoadMore,
  isLastPage,
}) => (
  <div className={cn(classes.productCardList, className)}>
    {products.map(product => (
      <ProductCard
        key={product.id}
        className={classes.productCard}
        toggleIsWishlisted={toggleIsWishlisted}
        {...product}
      />
    ))}
    <div className={classes.loadMoreButtonRoot}>
      {isLastPage === false && (
        <Button
          className={classes.loadMoreButton}
          onClick={handleClickLoadMore}
        >
          Load more
        </Button>
      )}
    </div>
  </div>
)

export const propTypes = {
  /** @ignore */
  classes: PropTypes.object.isRequired,
  /** Array of products to list */
  products: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      imgUrl: PropTypes.string.isRequired,
      gender: PropTypes.string.isRequired,
      rating: PropTypes.number.isRequired,
      price: PropTypes.number.isRequired,
      isNew: PropTypes.bool.isRequired,
      name: PropTypes.string.isRequired,
      brand: PropTypes.string.isRequired,
    }),
  ).isRequired,
  /**
   * Function to add an item to a wishlist
   * @param {id} string The id of the item to toggle the wishlist state
   */
  toggleIsWishlisted: PropTypes.func.isRequired,
  /** Gets called when the user clicks on the Load more button */
  handleClickLoadMore: PropTypes.func.isRequired,
  /** Property that defines if there are more products to load or not  */
  isLastPage: PropTypes.bool,
  /** Apply that className to the root component */
  className: PropTypes.string,
}

export const defaultProps = {
  isLastPage: null,
  className: null,
}

ProductCardList.propTypes = propTypes
ProductCardList.defaultProps = defaultProps

export default injectSheet(styles)(ProductCardList)
