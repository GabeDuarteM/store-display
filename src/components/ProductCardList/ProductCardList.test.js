import React from 'react'

import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import ProductCardList, {
  styles,
  propTypes,
  defaultProps,
} from './ProductCardList'
import ProductCard from '../ProductCard'
import Button from '../Button'

const createWrapper = (func, optionalProps) =>
  func(
    <ProductCardList
      products={[
        {
          id: '5a51804f840a30ab47a631a2',
          imgUrl: '',
          gender: 'woman',
          rating: 10,
          price: 27.59,
          isNew: true,
          name: 'Candacechase Blackturner',
          brand: 'NIKE',
        },
      ]}
      handleClickLoadMore={() => {}}
      toggleIsWishlisted={() => {}}
      {...optionalProps}
    />,
  )

const createWrapperShallow = optionalProps =>
  createWrapper(shallow, optionalProps)

const createWrapperMount = optionalProps => createWrapper(mount, optionalProps)

describe('COMPONENT: <ProductCardList />', () => {
  it('should render without crashing with default props', () => {
    createWrapperMount()
  })

  it("should match component's snapshot", () => {
    const wrapper = createWrapperShallow()
    expect(toJson(wrapper.dive())).toMatchSnapshot()
  })

  it("should match component's snapshot with custom props", () => {
    const wrapper = createWrapperShallow({ isLastPage: false })
    expect(toJson(wrapper.dive())).toMatchSnapshot()
  })

  it("should match component's style snapshot", () => {
    expect(styles).toMatchSnapshot()
  })

  it("should match component's propTypes snapshot", () => {
    expect(propTypes).toMatchSnapshot()
  })

  it("should match component's defaultProps snapshot", () => {
    expect(defaultProps).toMatchSnapshot()
  })

  it('should render a ProductCard for each product passed', () => {
    const products = [
      {
        id: '5a518050e421e659c6e24c46',
        imgUrl: 'https://tru-spec-cdn.azureedge.net/media/1175/1060f.png',
        gender: 'man',
        rating: 10,
        price: 282.1,
        isNew: true,
        name: 'Perkinsjames Clinegalloway',
        brand: 'VERO MODA',
      },
      {
        id: '5a51805044d1030570bcab76',
        imgUrl: 'https://tru-spec-cdn.azureedge.net/media/1025/1436f.png',
        gender: 'woman',
        rating: 4,
        price: 296.55,
        isNew: false,
        name: 'Henriettagillespie Dellavargas',
        brand: 'NIKE',
      },
    ]
    const wrapper = createWrapperShallow({ products })
    const productCardLength = wrapper.dive().find(ProductCard).length
    expect(productCardLength).toBe(2)
  })

  it('should have a Load more Button when isLastPage is false', () => {
    const wrapper = createWrapperShallow({ isLastPage: false })
    const button = wrapper.dive().find(Button)
    expect(button.exists()).toBe(true)
    expect(
      button
        .dive()
        .dive()
        .text(),
    ).toBe('Load more')
  })

  it('should not have a Load more Button when isLastPage is true', () => {
    const wrapper = createWrapperShallow({ isLastPage: true })
    const button = wrapper.dive().find(Button)
    expect(button.exists()).toBe(false)
  })

  it('should call handleClickLoadMore when the button is clicked', () => {
    const handleClickLoadMore = jest.fn()
    const wrapper = createWrapperShallow({
      isLastPage: false,
      handleClickLoadMore,
    })
    wrapper
      .dive()
      .find(Button)
      .simulate('click')
    expect(handleClickLoadMore).toHaveBeenCalledTimes(1)
  })
})
