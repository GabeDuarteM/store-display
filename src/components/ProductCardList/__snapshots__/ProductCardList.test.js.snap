// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT: <ProductCardList /> should match component's defaultProps snapshot 1`] = `
Object {
  "className": null,
  "isLastPage": null,
}
`;

exports[`COMPONENT: <ProductCardList /> should match component's propTypes snapshot 1`] = `
Object {
  "className": [Function],
  "classes": [Function],
  "handleClickLoadMore": [Function],
  "isLastPage": [Function],
  "products": [Function],
  "toggleIsWishlisted": [Function],
}
`;

exports[`COMPONENT: <ProductCardList /> should match component's snapshot 1`] = `
<div
  className="ProductCardList-productCardList"
>
  <Jss(ProductCard)
    brand="NIKE"
    className="ProductCardList-productCard"
    currency="R$"
    gender="woman"
    id="5a51804f840a30ab47a631a2"
    imgUrl=""
    isNew={true}
    isWishlisted={false}
    key="5a51804f840a30ab47a631a2"
    name="Candacechase Blackturner"
    price={27.59}
    rating={10}
    toggleIsWishlisted={[Function]}
  />
  <div
    className="ProductCardList-loadMoreButtonRoot"
  />
</div>
`;

exports[`COMPONENT: <ProductCardList /> should match component's snapshot with custom props 1`] = `
<div
  className="ProductCardList-productCardList"
>
  <Jss(ProductCard)
    brand="NIKE"
    className="ProductCardList-productCard"
    currency="R$"
    gender="woman"
    id="5a51804f840a30ab47a631a2"
    imgUrl=""
    isNew={true}
    isWishlisted={false}
    key="5a51804f840a30ab47a631a2"
    name="Candacechase Blackturner"
    price={27.59}
    rating={10}
    toggleIsWishlisted={[Function]}
  />
  <div
    className="ProductCardList-loadMoreButtonRoot"
  >
    <Jss(Button)
      className="ProductCardList-loadMoreButton"
      disabled={false}
      onClick={[Function]}
      width={280}
    >
      Load more
    </Jss(Button)>
  </div>
</div>
`;

exports[`COMPONENT: <ProductCardList /> should match component's style snapshot 1`] = `
Object {
  "loadMoreButton": Object {
    "margin": 25,
  },
  "loadMoreButtonRoot": Object {
    "alignItems": "center",
    "display": "flex",
    "justifyContent": "center",
    "width": "100%",
  },
  "productCard": Object {
    "flex": "none",
    "margin": 15,
  },
  "productCardList": Object {
    "alignItems": "center",
    "display": "flex",
    "flexWrap": "wrap",
    "justifyContent": "center",
    "margin": "0 47px",
    "padding": "15px 0",
  },
}
`;
